### PR TITLE
fix: scipy rel_entr edge case

### PR DESCRIPTION
**Pull request by sh0416**
#21192 

The value for edge case in rel_entr is 0, but the implementation set them to q distribution.
https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.special.rel_entr.html

I've fixed it, but need to be checked.

I am new for contributing this repo, so it would be pleased for someone to check this is right.

### CODE REVIEW

**Comment by jakevdp** [comment]
Thanks for the PR! This looks right the right fix, but we should also add some test coverage. Could you also change these two tests to use `rand_some_zero` rather than `rand_positive`? https://github.com/google/jax/blob/b4f214533769074dcb43842b5216d2dac2386916/tests/lax_scipy_special_functions_test.py#L145-L148

**Comment by sh0416** [comment]
I've checked your suggestiion. As the rand_some_zero contains negative values, but the two functions should work with negative values, so it is ok to change them as you suggested. Thank you for guiding me. Is there anything I have to add or change the codebase?

**Comment by jakevdp** [comment]
Hi - so it looks like there are problems with autodiff tests when we include zero. I guess I should have anticipated that: autodiff is evaluated with respect to finite differences, which will break down around domain edges. Let's try something like this instead:
```python
    # test autodiff for kl_div & rel_entr with positive inputs:
    op_record("kl_div", 2, float_dtypes, jtu.rand_positive, True),
    op_record("rel_entr", 2, float_dtypes, jtu.rand_positive, True),
    # test without autodiff for inputs including zero:
    op_record("kl_div", 2, float_dtypes, jtu.rand_some_zero, False),
    op_record("rel_entr", 2, float_dtypes, jtu.rand_some_zero, False),
```
This gives us autodiff test coverage in the well-defined domain, and output test coverage in the corner cases.

**Comment by sh0416** [comment]
Thank for the advice. I am doing it as is.

**Comment by sh0416** [comment]
Test name is duplicated I will handle immediately.

**Comment by sh0416** [comment]
I've check that the new test cases are passed except `LaxScipySpcialFunctionsTest.testScipySpecialFun_ndtri_1x4_float32_ndtri` which is irrelevant to this PR. It might be resolved in github CI, I wish..
@jakevdp Could you rerun the CI?

File: tests/lax_scipy_special_functions_test.py [hunk 1]
```diff
@@ -172,13 +176,14 @@ def _GetArgsMaker(self, rng, shapes, dtypes):
...
         if isinstance(rec.dtypes, list) else itertools.product(*rec.dtypes)),
+      test_name=[rec.test_name],
     ))
     for rec in JAX_SPECIAL_FUNCTION_RECORDS
   ))
   @jax.numpy_rank_promotion('allow')  # This test explicitly exercises implicit rank promotion.
   @jax.numpy_dtype_promotion('standard')  # This test explicitly exercises dtype promotion
   def testScipySpecialFun(self, op, rng_factory, shapes, dtypes,
-                          test_autodiff, nondiff_argnums):
+                          test_autodiff, nondiff_argnums, test_name):
```
Comment (jakevdp) [inline tests/lax_scipy_special_functions_test.py:186]: Can you say more about why this change is necessary? It doesn't look like `test_name` is used in the function.

Comment (sh0416) [reply tests/lax_scipy_special_functions_test.py:186]: In my chain of thought, the `test_name` variable is needed for generating the actual test name in the `pretty_special_fun_name` (to distinguish the test name for the new test case). However, `pytest.parameterize` carries whole values to the decorated function, so I just add them even though they are not used in the function body. I couldn't figure out how to do it without adding the new function argument, but if it is, I will change my implementation.

Comment (jakevdp) [reply tests/lax_scipy_special_functions_test.py:186]: Ah I see - it looks like the `test_name` is not used anymore after we refactored these tests last year. I think this is the wrong approach.

Instead of adding more test cases, let's leave the previous test cases alone and add a new manual test for each at the bottom of the file. What do you think?

Comment (jakevdp) [reply tests/lax_scipy_special_functions_test.py:186]: Actually, another thought if you're willing to look into it: I think the semantics of `jax.scipy.special.xlogy` match what we'd need for `rel_entr`, and there's a custom JVP rule to correctly handle gradients near zero. I'd bet if we re-implemented `rel_entr` in terms of `xlogy`, we could make the existing test work with the expanded domain. Is that something you want to explore?

Comment (jakevdp) [reply tests/lax_scipy_special_functions_test.py:186]: (Alternatively, you could just limit your change to the fix: replacing `q` with `zero` and I'll follow-up on the autodiff & testing questions)

Comment (sh0416) [reply tests/lax_scipy_special_functions_test.py:186]: I will re-implement it with xlogy.

Comment (sh0416) [reply tests/lax_scipy_special_functions_test.py:186]: It seems like `xlogy` is tested only in the domain of positive values. I've tried to test `xlogy` with `jtl.rand_some_zero`, but it could not pass the test. I think the jvp rule shows some instability. I change the division operator with `lax.div` shows some stable results, but they failed on two test cases, named `LaxScipySpcialFunctionsTest::testScipySpecialFun_xlogy_3x4_3x4_float32_float32` and `LaxScipySpcialFunctionsTest::testScipySpecialFun_xlogy_3x1_2x1x4_float32_float32`.

I already implemented rel_entr with xlogy. So, could you imagine why the jvp rule of `xlogy` is not work on the two testcases? I couldn't figure it out..

Comment (sh0416) [reply tests/lax_scipy_special_functions_test.py:186]:
```
E           AssertionError:
E           Not equal to tolerance rtol=0.3, atol=0.003
E           JVP tangent
E           x and y -inf location mismatch:
E            x: array([[0.],
E                  [0.],
E                  [0.]], dtype=float32)
E            y: array([[  0.],
E                  [-inf],
E                  [  0.]], dtype=float32)
```
for `LaxScipySpcialFunctionsTest.testScipySpecialFun_xlogy_3x1_2x1x4_float32_float32`.

```
E           AssertionError:
E           Not equal to tolerance rtol=1.2, atol=0.012
E           JVP tangent
E           x and y nan location mismatch:
E            x: array([[nan, nan, nan, nan],
E                  [nan, nan, nan, nan],
E                  [nan, nan, nan, nan]], dtype=float32)
E            y: array([[-inf, -inf, -inf, -inf],
E                  [  0.,   0.,   0.,   0.],
E                  [  0.,   0.,   0.,   0.]], dtype=float32)
```
For `LaxScipySpcialFunctionsTest.testScipySpecialFun_xlogy_3x4_3x4_float32_float32`.

Comment (jakevdp) [reply tests/lax_scipy_special_functions_test.py]: OK, then I was probably wrong about `xlogy` being applicable here. Thanks for checking.

Comment (sh0416) [reply tests/lax_scipy_special_functions_test.py]: Let's sort out these things.

First, I checked the behavior of `scipy.special.xlogy`.
```
>>> from scipy.special import xlogy
>>> xlogy([[1.1],[0],[-1.1]], [[1.1,0,-1.1]])
array([[ 0.1048412,       -inf,        nan],
       [ 0.       ,  0.       ,  0.       ],
       [-0.1048412,        inf,        nan]])
```
It seems that the output is well defined when `y>=0` or `x==0` (not nan region).

Second, I think whether the gradient, e.g., jvp, tangent, etc, could be computed through numerical approach, (f(x+eps)-f(x-eps))/(2*eps).

I've checked the code for testing the difference between custom jvp and numerical difference, and find out that they does not consider the nan value (only consider inf) and it is natural as the gradient could not be defined with nan value.

https://github.com/google/jax/blob/6fe313c6a1282158dfb0dccce61e65942368667d/jax/_src/public_test_util.py#L211-L215

https://github.com/google/jax/blob/6fe313c6a1282158dfb0dccce61e65942368667d/jax/_src/public_test_util.py#L187-L188

https://github.com/google/jax/blob/6fe313c6a1282158dfb0dccce61e65942368667d/jax/_src/public_test_util.py#L169-L173

In this case, when we compute the jvp tangent at y=0, using numerical approach could not be valid as f(x-eps) would be nan.
 
Third, I think about a general approach to compute the numerical gradient at the boundary case and one possible solution is to use left gradient or right gradient, (f(x)-f(x-eps))/eps or (f(x+eps)-f(x))/eps where the terms in the numerator would be valid in this case.

However, I am scared of changing the `check_grad` in `public_test_util.py` as it might broken lots of existing testcases, so I have to think about other approaches..

Comment (jakevdp) [reply tests/lax_scipy_special_functions_test.py]: Right, I think I was wrong that `xlogy` has relevant semantics. It doesn't and you should ignore that suggestion.

**Comment by jakevdp** [comment]
Hi - are you still interested in working on this? This is a bug that we really need to get fixed.

**Comment by sh0416** [comment]
I will sort it out within a hour. sorry for the late reply.

File: jax/_src/scipy/special.py [hunk 2]
```diff
@@ -665,14 +665,19 @@ def rel_entr(
...
   zero = _lax_const(p, 0.0)
-  both_gt_zero_mask = lax.bitwise_and(lax.gt(p, zero), lax.gt(q, zero))
-  one_zero_mask = lax.bitwise_and(lax.eq(p, zero), lax.ge(q, zero))
+
+  # if p is negative, set q to zero to produce output of xlogy to -inf
+  otherwise_mask = lax.bitwise_or(lax.lt(p, zero), lax.lt(q, zero))
+  otherwise_mask = lax.bitwise_or(otherwise_mask, lax.bitwise_and(lax.eq(q, zero), lax.ne(p, zero)))
+  safe_p = jnp.where(otherwise_mask, 1, p)
+  safe_q = jnp.where(otherwise_mask, 0, q)
+  return lax.sub(_xlogx(safe_p), xlogy(safe_p, safe_q))
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:674]: Please delete all this new code – it's not necessary as far as I can tell.

All we need is to change `q` to `zero`, as you've done below.

Comment (sh0416) [reply jax/_src/scipy/special.py]: I've deleted it.

**Comment by jakevdp** [comment]
Thanks, looks good! Last thing: can you please squash your changes into a single commit? (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests) This helps us keep the commit history clean and linear. Thanks!

**Comment by sh0416** [comment]
> Thanks, looks good! Last thing: can you please squash your changes into a single commit? (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests) This helps us keep the commit history clean and linear. Thanks!

I've done this. It seems clean. :)

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!
